crates/sui-framework/sources/crypto/randomness.move

```diff
@@ -0,0 +1,140 @@
+///   creation so even if the user aborts, depending on the randomness it received, the application
+///   is not harmed).
+///
+/// - Once set, the random value can be read/consumed.
+///
+///
+/// This object can be used as a shared-/owned-object.
+///
+module sui::randomness {
+    use std::hash::sha3_256;
+    use std::option;
+    use std::vector;
+    use sui::bcs;
+    use sui::object::{Self, UID, ID};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    /// Set is called with an invalid signature.
+    const EInvalidSignature: u64 = 0;
+    /// Already set object cannot be set again.
+    const EAlreadySet: u64 = 1;
+    /// Supplied randomness is not of the right length.
+    const EInvalidRndLength: u64 = 2;
+
+    /// All signatures are prefixed with Domain.
+    const Domain: vector<u8> = b"randomness";
+
+    struct Randomness<phantom T> has key {
+        id: UID,
+        epoch: u64,
+        value: option::Option<vector<u8>>
+    }
+
+    public fun new<T: drop>(_w: T, ctx: &mut TxContext): Randomness<T> {
+        Randomness<T> {
+            id: object::new(ctx),
+            epoch: tx_context::epoch(ctx),
+            value: option::none(),
+        }
+        // TODO: Front load the fee.
+    }
+
+    public fun transfer<T>(self: Randomness<T>, to: address) {
+        transfer::transfer(self, to);
+    }
+
+    public fun share_object<T>(self: Randomness<T>) {
+        transfer::share_object(self);
+    }
+
+    /// Owner(s) can use this function for setting the randomness.
+    public fun set<T>(self: &mut Randomness<T>, sig: vector<u8>) {
+        assert!(option::is_none(&self.value), EAlreadySet);
+        let msg = to_bytes(&Domain, self.epoch, &object::id(self));
+        assert!(native_tbls_verify_signature(self.epoch, &msg, &sig), EInvalidSignature);
+        let hashed = sha3_256(sig);
+        self.value = option::some(hashed);
+    }
+
+    /// Delete the object.
+    public fun destroy<T>(r: Randomness<T>) {
+        let Randomness { id, epoch: _, value: _ } = r;
+        object::delete(id);
+    }
+
+    /// Read the epoch of the object.
+    public fun epoch<T>(self: &Randomness<T>): u64 {
+        self.epoch
+    }
+
+    /// Read the current value of the object.
+    public fun value<T>(self: &Randomness<T>): &option::Option<vector<u8>> {
+        &self.value
+    }
+
+    fun to_bytes(domain: &vector<u8>, epoch: u64, id: &ID): vector<u8> {
+        let buffer: vector<u8> = vector::empty();
+        // All elements below are of fixed sizes.
+        vector::append(&mut buffer, *domain);
+        vector::append(&mut buffer, bcs::to_bytes(&epoch));
+        vector::append(&mut buffer, object::id_to_bytes(id));
+        buffer
+    }
+
+    /// Verify signature sig on message msg using the epoch's BLS key.
+    native fun native_tbls_verify_signature(epoch: u64, msg: &vector<u8>, sig: &vector<u8>): bool;
```

Comment:
Since these functions are relatively important (although not available publicly, it might be better to move them up for visibility).